File: src/analysis/hiring.cpp
// hiring.cpp — CLRS 4e Ch5 hiring problem 模擬：
// 隨機排列下期望雇用次數 = H_n ≈ ln n + 0.577
#include <algorithm>
#include <iostream>
#include <numeric>
#include <random>

int hireCount(std::vector<int>& rank, std::mt19937& rng) {
    std::shuffle(rank.begin(), rank.end(), rng);   // RANDOMIZED-HIRE-ASSISTANT 的第一步
    int best = -1, hires = 0;
    for (int r : rank)
        if (r > best) { best = r; ++hires; }       // 比目前最佳者好 → 雇用
    return hires;
}

int main() {
    const int n = 100, trials = 100000;
    std::vector<int> rank(n);
    std::iota(rank.begin(), rank.end(), 1);
    std::mt19937 rng(42);
    long long total = 0;
    for (int t = 0; t < trials; ++t) total += hireCount(rank, rng);
    double Hn = 0;
    for (int k = 1; k <= n; ++k) Hn += 1.0 / k;
    std::cout << "模擬平均雇用次數 = " << static_cast<double>(total) / trials
              << "（理論值 H_" << n << " = " << Hn << "）\n";
    return 0;
}

File: src/analysis/hiring.py
# hiring problem 模擬：RANDOMLY-PERMUTE（Fisher–Yates，同官方 Ch5 版）＋雇用計數
import random

random.seed(42)                             # 固定 seed 讓輸出可重現


def randomly_permute(A, n):
    for i in range(n):
        j = random.randint(i, n - 1)        # 從 A[i..n-1] 均勻挑一個和 A[i] 交換
        A[i], A[j] = A[j], A[i]


def hire_count(n):
    A = list(range(1, n + 1))               # rank 1..n
    randomly_permute(A, n)
    best = hires = 0
    for r in A:
        if r > best:                        # 比目前最佳好 → 換人（雇用）
            best, hires = r, hires + 1
    return hires


n, trials = 100, 20000
avg = sum(hire_count(n) for _ in range(trials)) / trials
H = sum(1 / i for i in range(1, n + 1))
print("模擬平均雇用次數 = %.4f（理論值 H_100 = %.4f）" % (avg, H))
